Add red-black tree implementation backed by the C version.

wedsonaf opened the pull request:
Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

ksquirrel commented:
Review of `1898079a9aa2`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`1898079`](https://github.com/Rust-for-Linux/linux/commit/1898079a9aa2f57aeb32f4124b62ad1d255f58ff): Looks fine!

<!-- Review 1898079a9aa2f57aeb32f4124b62ad1d255f58ff -->

ojeda commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,493 @@
+    /// Inserts a new node into the tree.
+    ///
+    /// It overwrites a node if one already exists with the same key and returns it (containing the
+    /// key/value pair). Returns `None` if a node with the same key didn't already exist.
```
In a few places we can link `None`:

```suggestion
    /// key/value pair). Returns [`None`] if a node with the same key didn't already exist.
```

wedsonaf replied:
Fixed here and in a few other places.

bjorn3 commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,493 @@
+
+        // SAFETY: All pointers are valid. `node` has just been inserted into the tree.
+        unsafe { bindings::rb_insert_color(&mut node.links, &mut self.root) };
+        Box::into_raw(node);
```
This invalidates all pointers to `node`. You could use `let node = ManuallyDrop::new(node);` or `let node = Box::into_raw(node);` before taking any pointers to it.

wedsonaf replied:
Thanks for noticing this. Updated the code to call `into_raw` in the beginning.

ojeda commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,493 @@
+///
+///     // Print all elements again.
+///     for (key, value) in &tree {
+///         pr_info!("{} = {}\n", key, value);
```
Thanks ***a lot*** for taking the time to write the examples.

One thing: if we use `assert!` and friends instead of printing, then it is even better, because it doubles as a test and readers also see the expected values. Since rbtrees are sorted, I assume it is possible to test for equality via iteration, either advancing manually or comparing against e.g. `30 * i` keeping the count. (If it does not look nice, I would still show the `for` notation via printing nevertheless!).

wedsonaf replied:
I added checks for the contents of the tree.

bjorn3 commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,493 @@
+        // SAFETY: `root` is valid as it's embedded in `self` and we have a valid `self`.
+        let mut next = unsafe { bindings::rb_first_postorder(&self.root) };
+
+        // INVARIANT: The loop invariant is that all tree nodes from `next` in postoder are valid.
```
*postorder

wedsonaf replied:
Fixed.

ojeda reviewed (commented):
Looks very nice.

A couple nits that can be skipped if needed.

bjorn3 commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,493 @@
+                    // SAFETY: All pointers are non-null and valid (parent, despite the name, really
+                    // is the node we're replacing).
+                    unsafe { bindings::rb_replace_node(parent, &mut node.links, &mut self.root) };
+                    Box::into_raw(node);
```
Same invalidation problem here.

wedsonaf replied:
Thanks. Fixed here too.

ksquirrel commented:
Review of `6e35e48f2440`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`6e35e48`](https://github.com/Rust-for-Linux/linux/commit/6e35e48f2440995acaffc6ec02f988e4ae63be8c): Looks fine!

<!-- Review 6e35e48f2440995acaffc6ec02f988e4ae63be8c -->

wedsonaf commented:
### v1 -> v2

* Updated samples to assert the contents of the tree.
* Updated references to `None` in documentation.
* Fixed typo in postorder comment.
* Fixed node pointer invalidation.

bjorn3 commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,540 @@
+        unsafe { rust_helper_rb_link_node(&mut node.links, parent, new_link) };
+
+        // SAFETY: All pointers are valid. `node` has just been inserted into the tree.
+        unsafe { bindings::rb_insert_color(&mut node.links, &mut self.root) };
```
The mutable reference to `node.links` is invalidated after this function call, but the function stores it in `self.root`, right? I think you should use `core::ptr::addr_of_mut!(node.links)` instead. Note that the `let node = unsafe { &mut *Box::into_raw(node) };` is also invalidated once this function returns. So maybe `let node = Box::into_raw(node);` and then `addr_of_mut!((*node).links)`? The `&mut self.root` is fine if there are no pointers to `self.root` only pointers stored in `self.root`.

wedsonaf replied:
I think I need to update my mental model here.

What is invalidating `node.links`? How would using `addr_of_mut!(node.links)` avoid this invalidation?

bjorn3 replied:
Thinking a bit more about this, this `&mut node.links` doesn't get invalidated, but the previous `&mut node.links` does due to this `&mut node.links`. `addr_of_mut!(node.links)` avoids this problem by directly creating a raw pointer without an intermediate mutable reference. It is valid to have multiple raw pointers to the same value, but creating one mutable reference invalidates all other references and pointers.

wedsonaf replied:
What model are you using to determine what gets invalidated?

I quickly read the model in the [stacked borrows](https://plv.mpi-sws.org/rustbelt/stacked-borrows/paper.pdf) paper and my understanding is that the code is ok now. The reason is that the borrow stack for each byte that makes up `node` contains: `[..., SharedRW, Unique(n), ...]`, where `SharedRW` refers to the raw pointer returned by `into_raw()` and `Unique(n)` refers to the mutable reference created on line 248. So any new raw references that I create in the function can still be used after the function returns because they will be resolved by the `SharedRW` entry (until I call `from_raw()`, which will pop that `SharedRW` entry) because raw pointers are not "tagged".

IOW, I don't agree with your claim that "creating one mutable reference invalidates all other pointers" -- my understanding is that it invalidates all other pointers on top of it in the borrow stack. Pointers below it continue to be valid until something else pops them off.

I am new to stacked borrows though, so my understanding may be wrong. Where does it not match yours?

bjorn3 replied:
> I quickly read the model in the stacked borrows paper and my understanding is that the code is ok now. The reason is that the borrow stack for each byte that makes up node contains: [..., SharedRW, Unique(n), ...], where SharedRW refers to the raw pointer returned by into_raw() and Unique(n) refers to the mutable reference created on line 248. So any new raw references that I create in the function can still be used after the function returns because they will be resolved by the SharedRW entry (until I call from_raw(), which will pop that SharedRW entry) because raw pointers are not "tagged".

You will end up with `SharedRW(k), Unique(l), SharedRW(m)` in this case. It  is fine with the default options of miri as the raw pointer can resolve to either `SharedRW`, but it isn't when `-Zmiri-track-raw-pointers` as it will only resolve to `SharedRW(m)` which is popped of the borrow stack at the next `&mut item.links` in addition `SharedRW(m)` only covers `links` and not the whole item, so accessing the rest of the item through it is not allowed.

> IOW, I don't agree with your claim that "creating one mutable reference invalidates all other pointers" -- my understanding is that it invalidates all other pointers on top of it in the borrow stack. Pointers below it continue to be valid until something else pops them off.

Yeah, I was simplifying a bit. Maybe a bit too much.

wedsonaf replied:
> You will end up with `SharedRW(k), Unique(l), SharedRW(m)` in this case. It is fine with the default options of miri as the raw pointer can resolve to either `SharedRW`, but it isn't when `-Zmiri-track-raw-pointers` as it will only resolve to `SharedRW(m)` which is popped of the borrow stack at the next `&mut item.links` in addition `SharedRW(m)` only covers `links` and not the whole item, so accessing the rest of the item through it is not allowed.

From miri's [readme](https://github.com/rust-lang/miri/blob/master/README.md) about `-Zmiri-track-raw-pointers`: _This can make valid code fail to pass the checks_. So I'm not so concerned about it, I'm concerned about the code correctness wrt the model. Do we agree that it is correct (wrt the model)?

wedsonaf replied:
Part of the reason I declared a `&mut Node` was to avoid having to add more `SAFETY` justifications in each of its uses (4 other places), but now I realise that 3 of them are `&mut node.links` so I can avoid repeating the justification by storing a raw pointer in a variable and reusing it.

I'll make this change and upload it shortly.

bjorn3 replied:
The new version is UB free I think.

bjorn3 commented on rust/kernel/rbtree.rs:
```diff
@@ -0,0 +1,540 @@
+        RBTreeNode {
+            // SAFETY: The pointer came from a `MaybeUninit<Node>` whose fields have all been
+            // initialised. Additionally, it has the same layout as `Node`.
+            node: unsafe { Box::from_raw(raw as _) },
```
Not a correctness issue, but you can use `Box::assume_init` on the `Box<MaybeUninit<Node>>` on nightly (the `new_uninit` feature). Please either use it here or add a FIXME for once the `new_uninit` feature is stabilized.

wedsonaf replied:
Yes, I could, but I don't want to use yet another unstable feature.

I also don't want to pollute the code with a FIXME because there is nothing to fix yet. I've opened issue #401 so that we don't lose track of this though.

ksquirrel commented:
Review of `fbe2d3444ab2`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`fbe2d34`](https://github.com/Rust-for-Linux/linux/commit/fbe2d3444ab2d510947ffe6f5f86780c7190a61c): Looks fine!

<!-- Review fbe2d3444ab2d510947ffe6f5f86780c7190a61c -->

wedsonaf commented:
### v2 -> v3

* Added `keys`, `values`, and `values_mut` per @nbdd0121's suggestion.
* Updated sample to use `values_mut`.
* Updated samples to assert the end of iteration (to possibly catch bugs when an extra node is left over).

ksquirrel commented:
Review of `3444d9b72b1b`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`3444d9b`](https://github.com/Rust-for-Linux/linux/commit/3444d9b72b1bcb02b5dd688264c2582ca6b23910): Looks fine!

<!-- Review 3444d9b72b1bcb02b5dd688264c2582ca6b23910 -->

wedsonaf commented:
### v3 -> v4

* Rebased
* Derive raw pointers from the one returned by `Box::into_raw`
